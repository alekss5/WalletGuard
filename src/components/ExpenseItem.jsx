import { View, Text, StyleSheet, Image, TouchableOpacity } from 'react-native';
import { useTheme } from 'react-native-paper';
import AntDesign from '@expo/vector-icons/AntDesign';
import ImageSource from '../utils/ImageSources';
import CustomDivider from './UI/CustomDivider';
import CustomText from './UI/CustomText';

export default function ExpenseItem({ date, title, currency, amount, onDelete,dayNumber }) {
  const { colors } = useTheme();

  return (
    <View>
      <View style={styles.expenseItem}>
        <View style={styles.itemLeft}>
          {dayNumber && <CustomText style={styles.currencyText} >{dayNumber}</CustomText>}
          <Image source={ImageSource.getImageSource(title.icon)} style={styles.icon} testID="expense-icon" />
          <Text style={[styles.itemText, { color: colors.text }]}>{title.description}</Text>
        </View>

        <View style={styles.itemRight}>
          <View>
          {date && <Text style={{ color: colors.subtext, fontSize: 15,paddingRight:'5%' }}>{date}</Text>}

          </View>
          
          <Text style={[styles.currencyText, { color: colors.text }]}>{currency}</Text>
          <Text style={[styles.amountText, { color: colors.text }]}>{amount}</Text>
          {onDelete && (
            <TouchableOpacity onPress={onDelete} style={{ paddingLeft: 10 }} testID="delete-button">
              <AntDesign name="delete" size={27} color={colors.text} />
            </TouchableOpacity>
          )}
        </View>
      </View>
      <CustomDivider />
    </View>
  );
}

const styles = StyleSheet.create({
  expenseItem: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    width: '90%',
    alignSelf: 'center',
    paddingBottom: 2,
  },
  itemLeft: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  icon: {
    width: 35,
    height: 35,
  },
  itemText: {
    fontSize: 18,
    paddingHorizontal: 10,
  },
  itemRight: {
    flexDirection: 'row',
    alignItems: 'center',
  },
  currencyText: {
    fontSize: 18,
  },
  amountText: {
    fontSize: 18,
    paddingLeft: 5,
  },
});
