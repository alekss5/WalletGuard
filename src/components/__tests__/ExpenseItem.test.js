import React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import ExpenseItem from '../ExpenseItem';
import { ThemeProvider } from 'react-native-paper';

const mockOnDelete = jest.fn();

const theme = {
  colors: {
    text: '#000',
    subtext: '#666',
  }
};

describe('ExpenseItem Component', () => {

  // Renders component with all props
  it('renders the ExpenseItem component correctly with all props', () => {
    const { getByText, getByTestId } = render(
      <ThemeProvider theme={theme}>
        <ExpenseItem
          date="2024-09-30"
          title={{ description: 'Shopping', icon: 'shopping' }}
          currency="USD"
          amount="100.00"
          onDelete={mockOnDelete}
          dayNumber="30"
        />
      </ThemeProvider>
    );

    // Verify the title is displayed
    expect(getByText('Shopping')).toBeTruthy();

    // Verify the date is displayed
    expect(getByText('2024-09-30')).toBeTruthy();

    // Verify the currency and amount are displayed
    expect(getByText('USD')).toBeTruthy();
    expect(getByText('100.00')).toBeTruthy();

    // Find the delete button by testID
    const deleteButton = getByTestId('delete-button');
    expect(deleteButton).toBeTruthy();

    // Test delete button click
    fireEvent.press(deleteButton);
    expect(mockOnDelete).toHaveBeenCalled();
  });

  // Renders without the delete button when onDelete is not provided
  it('renders the component without the delete button when onDelete is not provided', () => {
    const { queryByTestId } = render(
      <ThemeProvider theme={theme}>
        <ExpenseItem
          date="2024-09-30"
          title={{ description: 'Shopping', icon: 'shopping' }}
          currency="USD"
          amount="100.00"
          dayNumber="30"
        />
      </ThemeProvider>
    );

    // The delete button should not be rendered when onDelete is not provided
    expect(queryByTestId('delete-button')).toBeNull();
  });

  // Test conditional rendering of dayNumber
  it('does not render dayNumber when it is not provided', () => {
    const { queryByText } = render(
      <ThemeProvider theme={theme}>
        <ExpenseItem
          title={{ description: 'Shopping', icon: 'shopping' }}
          currency="USD"
          amount="100.00"
        />
      </ThemeProvider>
    );
    
    // dayNumber should not be rendered
    expect(queryByText('30')).toBeNull();
  });

  // Test conditional rendering of date
  it('does not render date when it is not provided', () => {
    const { queryByText } = render(
      <ThemeProvider theme={theme}>
        <ExpenseItem
          title={{ description: 'Shopping', icon: 'shopping' }}
          currency="USD"
          amount="100.00"
          dayNumber="30"
        />
      </ThemeProvider>
    );
    
    // date should not be rendered
    expect(queryByText('2024-09-30')).toBeNull();
  });

  // Test icon rendering based on the title
  it('renders the correct icon based on the title', () => {
    const { getByTestId } = render(
      <ThemeProvider theme={theme}>
        <ExpenseItem
          title={{ description: 'Shopping', icon: 'shopping' }}
          currency="USD"
          amount="100.00"
        />
      </ThemeProvider>
    );
  
    // Query by testID
    const icon = getByTestId('expense-icon');
    expect(icon).toBeTruthy(); // Check if the image exists
  });
  

  // Snapshot test
  it('matches the snapshot', () => {
    const tree = renderer.create(
      <ThemeProvider theme={theme}>
        <ExpenseItem
          date="2024-09-30"
          title={{ description: 'Shopping', icon: 'shopping' }}
          currency="USD"
          amount="100.00"
          onDelete={mockOnDelete}
          dayNumber="30"
        />
      </ThemeProvider>
    ).toJSON();
    
    expect(tree).toMatchSnapshot();
  });
});
