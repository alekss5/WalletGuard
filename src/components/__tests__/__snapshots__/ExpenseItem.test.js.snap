// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseItem Component matches the snapshot 1`] = `
<View>
  <View
    style={
      {
        "alignSelf": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingBottom": 2,
        "width": "90%",
      }
    }
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <Text
        style={
          [
            {
              "color": "#000",
            },
            {
              "fontSize": 18,
            },
          ]
        }
      >
        30
      </Text>
      <Image
        source={null}
        style={
          {
            "height": 35,
            "width": 35,
          }
        }
        testID="expense-icon"
      />
      <Text
        style={
          [
            {
              "fontSize": 18,
              "paddingHorizontal": 10,
            },
            {
              "color": "#000",
            },
          ]
        }
      >
        Shopping
      </Text>
    </View>
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
        }
      }
    >
      <View>
        <Text
          style={
            {
              "color": "#666",
              "fontSize": 15,
              "paddingRight": "5%",
            }
          }
        >
          2024-09-30
        </Text>
      </View>
      <Text
        style={
          [
            {
              "fontSize": 18,
            },
            {
              "color": "#000",
            },
          ]
        }
      >
        USD
      </Text>
      <Text
        style={
          [
            {
              "fontSize": 18,
              "paddingLeft": 5,
            },
            {
              "color": "#000",
            },
          ]
        }
      >
        100.00
      </Text>
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
            "paddingLeft": 10,
          }
        }
        testID="delete-button"
      >
        <AntDesign
          color="#000"
          name="delete"
          size={27}
        />
      </View>
    </View>
  </View>
  <View
    style={
      [
        {
          "backgroundColor": "rgba(0, 0, 0, 0.12)",
          "height": 0.5,
        },
        undefined,
        undefined,
        undefined,
        {
          "alignSelf": "center",
          "backgroundColor": "#ccc",
          "height": 2,
          "marginBottom": 10,
          "marginVertical": 10,
          "width": "90%",
        },
      ]
    }
  />
</View>
`;
